Warn users when `TargetFramework` and `RuntimeIdentifier` are not lowercase

Nirmal4G opened the pull request:
#### CHANGES

Instead of Implicitly converting to lowercase, warn users instead. So, when possible, they can change them to lowercase right at the source.
We don't actually require lowercase but it was assumed in the design doc since there is logic when the values are used as a part of the path in case-sensitive file systems.

#### NOTES

The comments and the help text are not final, since, I'm not fluent in English.
Please do suggest alternative forms that properly convey the meaning and intent.
I'll update the tests and localizations after we land on something more concrete.

Nirmal4G committed:
Warn users when 'TargetFramework'/'RuntimeIdentifier' is not lowercase

It is recommended to use only lower-casing for the above properties.
Technically, any project should build with any casing but when dealing
with case-sensitive file-systems, it gets complicated. MSBuild properties
are case-insensitive but they do preserve case. This leads to different paths
in an case-sensitive file-system.

To ensure we have same paths everywhere, we could make them lower-case before
appending to the path string. But this raises issues when these properties
are set programmatically in a script and using them without conversion
will lead to file not found errors. This also incurs some perf.

Instead of an implicit conversion—which also removes the choice to have
different casing when needed, warn about the casing of these values.

Nirmal4G committed:
Remove implicit lower-casing of 'TargetFramework'/'RuntimeIdentifier'

Lowercase is recommended for the above properties but implicitly converting
them to lowercase removes the choice to have different casing when needed.
Calls to 'ToLower' also incur some perf cost too.

Lower-casing is needed only when the output paths appends these properties,
and that too in a case-sensitive file-system. This also raises some issues
when these properties are set programmatically in a script and using them
without conversion will lead to file not found errors.

Instead of an implicit conversion, let's keep what is specified and probably
a warning about the casing might be an user friendly solution.

Nirmal4G commented:
@marcpopMSFT @sfoslund

Nirmal4G commented:
@marcpopMSFT @sfoslund 🔔

sfoslund commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:
```diff
@@ -185,6 +185,19 @@ Copyright (c) .NET Foundation. All rights reserved.
+          BeforeTargets="_CheckForInvalidConfigurationAndPlatform;RunResolvePackageDependencies;GetFrameworkPaths;GetReferenceAssemblyPaths;Restore"
+          Condition="'$(AppendRuntimeIdentifierToOutputPath)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(_UsingDefaultRuntimeIdentifier)' != 'true'">
+
+    <NETSdkError Condition="!$(RuntimeIdentifier.Equals('$(RuntimeIdentifier.ToLowerInvariant())'))"
```
If we include this I think it should be a warning, not a error.

Nirmal4G replied:
Yeah!! My mistake.

ViktorHofer commented:
This will break dotnet/runtime as we use custom TargetFramework values and some of those aren't entirely lowercase. Means we likely need an escape hatch. cc @Anipik @ericstj 

Aside from that, is the validation worth the cost?

Nirmal4G commented:
See that it only validates, when the values are used in the path only when our default paths logic takes effect.

IMHO, the validation is worth it. It gives out a nice warning when you are leaving the paths to be default from the SDK targets but using short TFMs which might not be in lower-case.

wli3 commented:
most of the msbuild properties are case insensitive. I don't think we should single out these 2

rainersigwald commented:
> Would we go through and remove the ToLower calls in msbuild if we decided to take this change?

I think that would probably be a user-project-breaking change so "no". As a result, I think this warning might add noise while not providing much benefit, so doesn't seem like a great idea overall to me. It's another in the "we probably should have done this all along but it's too late now" category.

richlander commented:
I don't have a strong opinion on the actual change.

> I think that would probably be a user-project-breaking change so "no".

Doesn't think call for a warning/error wave model? If we had one, then this particular concern wouldn't be as strong, right?

rainersigwald commented:
> Doesn't think call for a warning/error wave model? If we had one, then this particular concern wouldn't be as strong, right?

We do [have one in MSBuild](https://github.com/dotnet/msbuild/blob/main/documentation/wiki/ChangeWaves.md).

My speculation on this particular issue is that even with a changewave it's not a great path forward because of the small user benefit (remove some calls to `ToLower` for a small perf improvement). Even with changewaves I don't want to force user projects to churn unless there's fairly strong motivation.

richlander commented:
@rainersigwald -- Ah, I was missing the forest for the trees on that one. I agree with you. Thanks for explaining.

Anipik commented:
> This will break dotnet/runtime as we use custom TargetFramework values and some of those aren't entirely lowercase. Means we likely need an escape hatch. cc @Anipik @ericstj

we might want to just get rid of the lowercase stuff, we didnt do it in the past bcoz there was no extra benefit of doing this

ViktorHofer commented:
@Anipik you mean the uppercase stuff, not lower case, right?

Anipik commented:
> @Anipik you mean the uppercase stuff, not lower case, right?

correct

Nirmal4G committed:
Fix casing test for 'RuntimeIdentifier'

With the RID casing change, we don't expect it to be lower-case anymore.
But the change requires updates across NuGet, .NET SDK and VS IDE.
For now, we let the error pass through along with the warning.

Nirmal4G commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -237,6 +237,10 @@
     <value>NETSDK1032: The RuntimeIdentifier platform '{0}' and the PlatformTarget '{1}' must be compatible.</value>
     <comment>{StrBegin="NETSDK1032: "}</comment>
   </data>
+  <data name="PropertyValueShouldBeLowercase" xml:space="preserve">
+    <value>NETSDK1033: The value of '{0}' is expected to be in lower-case but the actual casing is '{1}'. This value might be used as a part of a path which could lead into unexpected file not found errors (i.e. in case-sensitive file-systems). So, please use lower-case for '{0}' value.</value>
```
The main motivation is, when a team of devs do the following… Then, for the guy (**ME**) who stuck at figuring out what's wrong with the build, this message would've been helpful to debug and fix the build, instead of banging my head against the wall, in a race against time!

Let's say, we could set TF and RID programmatically…

```proj
  <PropertyGroup>
    <TargetFramework Condition="'$(TargetFramework)' == ''">net3.1</TargetFramework>
    <RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">win-x64</RuntimeIdentifier>
  </PropertyGroup>
```

And a build script that does the following…

```ps1

  # 'tfx' and 'rid' could have uppercase chars
  tfx = get_tfx() # get framework moniker
  rid = get_rid() # get runtime moniker

  msbuild -t:Build;Validate -p:TargetFramework=$tfx -p:RuntimeIdentifier=$rid

  path = make_path($tfx, $rid) # contruct path using 'tfx' and 'rid'

  # some logic to copy files or run tests against those files…
```

Then, the above script would fail, since, the file is not found in a case-sensitive file system.

So, to me, this is not noise, this is godsend!

Nirmal4G commented:
The main motivation is, when a team of devs do the following… Then, for the guy (**ME**) who stuck at figuring out what's wrong with the build, this message would've been helpful to debug and fix the build, instead of banging my head against the wall, in a race against time!

Let's say, we could set TF and RID programmatically…

```proj
  <PropertyGroup>
    <TargetFramework Condition="'$(TargetFramework)' == ''">net5.0</TargetFramework>
    <RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">win-x64</RuntimeIdentifier>
  </PropertyGroup>
```

And a build script that does the following…

```ps1

  # 'tfx' and 'rid' could have uppercase chars
  tfx = get_tfx() # get framework moniker
  rid = get_rid() # get runtime moniker

  msbuild -t:Build;Validate -p:TargetFramework=$tfx -p:RuntimeIdentifier=$rid

  path = make_path($tfx, $rid) # contruct path using 'tfx' and 'rid'

  # some logic to copy files or run tests against those files…
```

Then, the above script would fail, since, the file is not found in a case-sensitive file system.

So, to me, this is not noise, this is godsend!

marcpopMSFT commented:
Clearing out old PRs. At this time we won't take this but we are considering adding msbuild analyzers in the future to help people detect common patterns and this might fit there.
